e2e-tests/tests/plan-external-source.test.ts

```diff
@@ -0,0 +1,393 @@
+import test, { expect, type BrowserContext, type Page } from '@playwright/test';
+import { Constraints } from '../fixtures/Constraints.js';
+import { ExternalSources } from '../fixtures/ExternalSources.js';
+import { Models } from '../fixtures/Models.js';
+import { PanelNames, Plan } from '../fixtures/Plan.js';
+import { Plans } from '../fixtures/Plans.js';
+import { SchedulingConditions } from '../fixtures/SchedulingConditions.js';
+import { SchedulingGoals } from '../fixtures/SchedulingGoals.js';
+import { User } from '../fixtures/User.js';
+
+let constraints: Constraints;
+let context: BrowserContext;
+let externalSources: ExternalSources;
+let models: Models;
+let page: Page;
+let plan: Plan;
+let plans: Plans;
+let schedulingConditions: SchedulingConditions;
+let schedulingGoals: SchedulingGoals;
+let userA: User;
+let userB: User;
+const extendedTimeout = 5000;
+
+test.beforeEach(async () => {
+  await plan.goto(); // Refresh page to reset the view
+});
+
+test.beforeAll(async ({ baseURL, browser }) => {
+  context = await browser.newContext();
+  page = await context.newPage();
+
+  models = new Models(page);
+  plans = new Plans(page, models);
+  plans.endTime = '2022-011T00:00:00'; // Extend to cover the whole derivation group example
+  constraints = new Constraints(page);
+  schedulingConditions = new SchedulingConditions(page);
+  schedulingGoals = new SchedulingGoals(page);
+  plan = new Plan(page, plans, constraints, schedulingGoals, schedulingConditions);
+  externalSources = new ExternalSources(page);
+
+  await models.goto();
+  await models.createModel(baseURL);
+  await plans.goto();
+  await plans.createPlan();
+  await externalSources.goto();
+  await externalSources.uploadExternalSource();
+});
+
+test.afterAll(async () => {
+  await plans.goto();
+  await plans.deletePlan();
+  await models.goto();
+  await models.deleteModel();
+
+  await externalSources.goto();
+  await page.waitForTimeout(500); // Arbitrary wait timing, allows external sources table to populate
+  // Cleanup all test files that *may* have been uploaded
+  if (await page.getByRole('gridcell', { name: externalSources.externalSourceFileName }).first().isVisible()) {
```

Comment:
Ah, my bad! Must have misclicked this one as resolved instead of another - I think this visibility check is fine to be done before deleting any source, so moving it inside of `deleteSource` itself!